framework/quota/src/org/apache/cloudstack/quota/dao/QuotaBalanceDaoImpl.java

```diff
@@ -0,0 +1,201 @@
+        try {
+            SearchCriteria<QuotaBalanceVO> sc = createSearchCriteria();
+            if (accountId != null) {
+                sc.addAnd("accountId", SearchCriteria.Op.EQ, accountId);
+            }
+            if (domainId != null) {
+                sc.addAnd("domainId", SearchCriteria.Op.EQ, domainId);
+            }
+            if ((startDate != null) && (endDate != null) && startDate.before(endDate)) {
+                sc.addAnd("updatedOn", SearchCriteria.Op.BETWEEN, startDate, endDate);
+            } else {
+                return new ArrayList<QuotaBalanceVO>();
+            }
+            quotaUsageRecords = listBy(sc);
+            if (quotaUsageRecords.size() == 0) {
+                quotaUsageRecords.addAll(lastQuotaBalanceVO(accountId, domainId, startDate));
+            }
+        } finally {
+            txn.close();
+        }
+
+        TransactionLegacy.open(opendb).close();
+        return quotaUsageRecords;
+    }
+
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public List<QuotaBalanceVO> lastQuotaBalanceVO(final Long accountId, final Long domainId, final Date pivotDate) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+
+        TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.USAGE_DB);
+        List<QuotaBalanceVO> quotaUsageRecords = null;
+        List<QuotaBalanceVO> trimmedRecords = new ArrayList<QuotaBalanceVO>();
+        try {
+            Filter filter = new Filter(QuotaBalanceVO.class, "updatedOn", false, 0L, 100L);
+            // ASSUMPTION there will be less than 100 continuous credit
+            // transactions
+            SearchCriteria<QuotaBalanceVO> sc = createSearchCriteria();
+            if (accountId != null) {
+                sc.addAnd("accountId", SearchCriteria.Op.EQ, accountId);
+            }
+            if (domainId != null) {
+                sc.addAnd("domainId", SearchCriteria.Op.EQ, domainId);
+            }
+            if ((pivotDate != null)) {
+                sc.addAnd("updatedOn", SearchCriteria.Op.LTEQ, pivotDate);
+            }
+            quotaUsageRecords = search(sc, filter);
+
+            // get records before startDate to find start balance
+            for (Iterator<QuotaBalanceVO> it = quotaUsageRecords.iterator(); it.hasNext();) {
+                QuotaBalanceVO entry = it.next();
+                s_logger.info("findQuotaBalance Date=" + entry.getUpdatedOn().toGMTString() + " balance=" + entry.getCreditBalance() + " credit=" + entry.getCreditsId());
+                if (entry.getCreditsId() > 0) {
+                    trimmedRecords.add(entry);
+                } else {
+                    trimmedRecords.add(entry);
+                    break; // add only consecutive credit entries and last balance entry
+                }
+            }
+
+        } finally {
+            txn.close();
+        }
+
+        TransactionLegacy.open(opendb).close();
```

Comment:
See previous comments regarding apparently needless open/close of a database transaction.